lib/Target/DebugInfo/EmitHGLDD.cpp

```diff
@@ -79,26 +89,117 @@ static FileLineColLoc findBestLocation(Location loc, bool emitted) {
 //===----------------------------------------------------------------------===//
 
 namespace {
 
+/// An emitted type.
+struct EmittedType {
+  StringRef name;
+  SmallVector<int64_t, 1> packedDims;
+  SmallVector<int64_t, 1> unpackedDims;
+
+  EmittedType() = default;
+  EmittedType(StringRef name) : name(name) {}
+  EmittedType(Type type) {
+    while (type) {
+      type = hw::getCanonicalType(type);
+      if (auto inoutType = dyn_cast<hw::InOutType>(type)) {
+        type = inoutType.getElementType();
+        continue;
+      }
+      if (hw::isHWIntegerType(type)) {
+        name = "logic";
+        addPackedDim(hw::getBitWidth(type));
+      }
+      break;
+    }
+  }
+
+  void addPackedDim(int64_t dim) { packedDims.push_back(dim); }
+  void addUnpackedDim(int64_t dim) { unpackedDims.push_back(dim); }
+
+  operator bool() const { return !name.empty(); }
+
+  static JArray emitDims(ArrayRef<int64_t> dims, bool skipFirstLen1Dim) {
+    JArray json;
+    if (skipFirstLen1Dim && !dims.empty() && dims[0] == 1)
+      dims = dims.drop_front();
+    for (auto dim : llvm::reverse(dims)) {
+      json.push_back(dim - 1);
+      json.push_back(0);
+    }
+    return json;
+  }
+  JArray emitPackedDims() const { return emitDims(packedDims, true); }
+  JArray emitUnpackedDims() const { return emitDims(unpackedDims, false); }
+};
+
+/// An emitted expression and its type.
+struct EmittedExpr {
+  JValue expr = nullptr;
+  EmittedType type;
+  operator bool() const { return expr != nullptr && type; }
+};
+
+llvm::raw_ostream &operator<<(llvm::raw_ostream &os, const EmittedType &type) {
+  if (!type)
+    return os << "<null>";
+  os << type.name;
+  for (auto dim : type.packedDims)
+    os << '[' << dim << ']';
+  if (!type.unpackedDims.empty()) {
+    os << '$';
+    for (auto dim : type.unpackedDims)
+      os << '[' << dim << ']';
+  }
+  return os;
+}
+
+llvm::raw_ostream &operator<<(llvm::raw_ostream &os, const EmittedExpr &expr) {
+  if (!expr)
+    return os << "<null>";
+  return os << expr.expr << " : " << expr.type;
+}
+
 /// Contextual information for a single HGLDD file to be emitted.
 struct FileEmitter {
   const EmitHGLDDOptions *options = nullptr;
+  const hw::HWSymbolCache *symbolCache = nullptr;
   SmallVector<DIModule *> modules;
   SmallString<64> outputFileName;
   StringAttr hdlFile;
   SmallMapVector<StringAttr, unsigned, 8> sourceFiles;
+  Namespace objectNamespace;
+  SmallMapVector<JValue, StringRef, 8> structDefs;
+  SmallString<128> structNameHint;
 
   void emit(llvm::raw_ostream &os);
-  void emit(llvm::json::OStream &json);
-  void emitLoc(llvm::json::OStream &json, FileLineColLoc loc, bool emitted);
-  void emitModule(llvm::json::OStream &json, DIModule *module);
-  void emitInstance(llvm::json::OStream &json, DIInstance *instance);
-  void emitVariable(llvm::json::OStream &json, DIVariable *variable);
+  void emit(JOStream &json);
+  JValue emitLoc(FileLineColLoc loc, FileLineColLoc endLoc, bool emitted);
```

Comment:
nit: Is it needed to have this return the JValue rather than taking a json stream by reference as input? It just stood out in this change as being slightly different from the other methods. Feel free to disregard this comment, just something I noticed.

Reply:
`emitLoc` is different than the other `emit` functions because it is used within `emitExpression` to build up the HGLDD expression objects, which aren't directly emitted to the stream but built up as `JObject`s first.